src/element/linearElementEditor.ts

```diff
@@ -0,0 +1,409 @@
+import {
+  NonDeleted,
+  ExcalidrawLinearElement,
+  ExcalidrawElement,
+} from "./types";
+import { distance2d, rotate, isPathALoop } from "../math";
+import { getElementAbsoluteCoords } from ".";
+import { getElementPointsCoords } from "./bounds";
+import { Point, AppState } from "../types";
+import { mutateElement } from "./mutateElement";
+import { SceneHistory } from "../history";
+import { globalSceneState } from "../scene";
+
+export class LinearElementEditor {
+  public elementId: ExcalidrawElement["id"];
+  public activePointIndex: number | null;
+  public draggingElementPointIndex: number | null;
+  public lastUncommittedPoint: Point | null;
+
+  constructor(element: NonDeleted<ExcalidrawLinearElement>) {
+    LinearElementEditor.normalizePoints(element);
+
+    this.elementId = element.id;
+    this.activePointIndex = null;
+    this.lastUncommittedPoint = null;
+    this.draggingElementPointIndex = null;
+  }
+
+  // ---------------------------------------------------------------------------
+  // static methods
+  // ---------------------------------------------------------------------------
+
+  static POINT_HANDLE_SIZE = 20;
+
+  static getElement(id: ExcalidrawElement["id"]) {
+    const element = globalSceneState.getNonDeletedElement(id);
+    if (element) {
+      return element as NonDeleted<ExcalidrawLinearElement>;
+    }
+    return null;
+  }
+
+  /** @returns whether point was dragged */
+  static handlePointDragging(
+    appState: AppState,
```

Comment:
would prefer for us to just pass the parts of the state that are needed, and a callback that just updates that part of the state rather than the full setState(). reason being that i want to stop passing appState around so we can implement `shouldComponentUpdate()` in our child components.

Reply:
> callback that just updates that part of the state rather than the full setState().

If you mean just returning the `editingLinearElement` object and setting the state upstream (as I'm doing elsewhere), then that was my initial idea, but it would leak too much logic outside to `App.tsx` (and complicate the return value of `handlePointDragging` etc). I wanted to keep the logic local to `linearElementEditors.ts` as much as possible. Not sure I can do both at the same time.

> reason being that i want to stop passing appState around so we can implement shouldComponentUpdate() in our child components.

Unclear how that relates.